In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/google-play-store-data/googleplaystore.csv


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#501ee6>About Author:</font></h2>

<a><span style='color:#501ee6; font-size:15px'>**Name :** |</span><span style='color:#01181c; font-size:15px'> Muhammad Adnan (Pakistan)</span></a> \
<a><span style='color:#501ee6; font-size:15px'>**Profession :** |</span><span style='color:#01181c; font-size:15px'>  Data Scientist</span></a>\
<a><span style='color:#501ee6; font-size:15px'>**Data :** |</span><span style='color:#01181c; font-size:15px'>  Google Play store</span></a>\
<a><span style='color:#501ee6; font-size:15px'>**Date :** |</span><span style='color:#01181c; font-size:15px'>  08-Novermber, 2023</span></a>

<img src='https://logos-world.net/wp-content/uploads/2020/11/Google-Play-Logo-2015-2016.png' width=400px>

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#cc0a27>About data:</font></h2>

<a><span style='color:#cc0a27; font-size:15px'>**App :** |</span><span style='color:#01181c; font-size:15px'> The name of the apps (Different)</span></a> \
<a><span style='color:#cc0a27; font-size:15px'>**Category :** |</span><span style='color:#01181c; font-size:15px'>  The category of the app (e.g; Education, Entertainment, Life style etc.</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Rating :** |</span><span style='color:#01181c; font-size:15px'>   The rating has a range from (0-5)</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Reviews :** |</span><span style='color:#01181c; font-size:15px'>  The number of reviews of the app</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Size :**  |</span><span style='color:#01181c; font-size:15px'>  The size of the app in (kb/Mb)</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Install :** |</span><span style='color:#01181c; font-size:15px'>   The number of installs of the app</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Type :**  |</span><span style='color:#01181c; font-size:15px'>  The type of the app (Free/Paid)</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Price :**  |</span><span style='color:#01181c; font-size:15px'>  The price of the app (Free/Paid)</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Content Rating :** |</span><span style='color:#01181c; font-size:15px'>  The different individual categories</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Genres:**  |</span><span style='color:#01181c; font-size:15px'>   The genre of the app</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Last Updated :**  |</span><span style='color:#01181c; font-size:15px'>  The date when the app was last updated</span></a>\
<a><span style='color:#cc0a27; font-size:15px'>**Current Ver :** |</span><span style='color:#01181c; font-size:15px'>  The current version of the app</span></a> \
<a><span style='color:#cc0a27; font-size:15px'>**Android Ver :**  |</span><span style='color:#01181c; font-size:15px'>  The minimum Android version required to run the app</span></a>

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#501ee6>Google Play Store:</font></h2>

<a><span style='color:#501ee6; font-size:15px'>**About the Data :** |</span><span style='color:#01181c; font-size:15px'> While many public datasets (on Kaggle and the like) provide Apple App Store data, there are not many counterpart datasets available for Google Play Store apps anywhere on the web. On digging deeper, I found out that iTunes App Store page deploys a nicely indexed appendix-like structure to allow for simple and easy web scraping. On the other hand, Google Play Store uses sophisticated modern-day techniques (like dynamic page load) using JQuery making scraping more challenging.</span></a> 

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h3 align="left"><font color=#501ee6>Data Handling steps:</font></h3>

<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>Data Loading</li>
     <li>Data Exploration</li>
     <li>Data Wrangling</li>
     <li>Data visualization</li>
    <li>Data questioning & Visualization</li></ul></span></a> 

<a id="load_dataset"></a>
# <b><span style='color:#cc0a27'>Step 1 |</span><span style='color:#0aaccc'> Loading the Libraries</span></b>

In [2]:
# Load the libraries
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
%matplotlib inline

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h3 align="left"><font color=#501ee6>Data Loading, Exploration & Cleaning :</font></h3>

<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>Load the csv file with the pandas</li>
     <li>Creating the dataframe and understanding the data present in the dataset using pandas</li>
     <li>Dealing with the missing data, outliers and the incorrect records</li>
     <li>Remove Duplicates rows</li></ul></span></a> 

In [3]:
df = pd.read_csv("/kaggle/input/google-play-store-data/googleplaystore.csv")
df.head(3)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   

   Reviews  Size    Installs  Type Price Content Rating  \
0      159   19M     10,000+  Free     0       Everyone   
1      967   14M    500,000+  Free     0       Everyone   
2    87510  8.7M  5,000,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up  
2               Art & Design    August 1, 2018       1.2.4  4.0.3 and up

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#cc0a27>Data Columns and Rows:</font></h2>

<a><span style='color:#cc0a27; font-size:15px'>**App :** |</span><span style='color:#01181c; font-size:15px'> The name of the apps (Different)</span></a> 
<a><span style='color:#cc0a27; font-size:15px'>**Category :** |</span><span style='color:#01181c; font-size:15px'>  The category of the app (e.g; Education, Entertainment, Life style etc.</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Rating :** |</span><span style='color:#01181c; font-size:15px'>   The rating has a range from (0-5)</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Reviews :** |</span><span style='color:#01181c; font-size:15px'>  The number of reviews of the app</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Size :**  |</span><span style='color:#01181c; font-size:15px'>  The size of the app in (kb/Mb)</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Install :** |</span><span style='color:#01181c; font-size:15px'>   The number of installs of the app</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Type :**  |</span><span style='color:#01181c; font-size:15px'>  The type of the app (Free/Paid)</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Price :**  |</span><span style='color:#01181c; font-size:15px'>  The price of the app (Free/Paid)</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Content Rating :** |</span><span style='color:#01181c; font-size:15px'>  The different individual categories</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Genres:**  |</span><span style='color:#01181c; font-size:15px'>   The genre of the app</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Last Updated :**  |</span><span style='color:#01181c; font-size:15px'>  The date when the app was last updated</span></a>
<a><span style='color:#cc0a27; font-size:15px'>**Current Ver :** |</span><span style='color:#01181c; font-size:15px'>  The current version of the app</span></a> 
<a><span style='color:#cc0a27; font-size:15px'>**Android Ver :**  |</span><span style='color:#01181c; font-size:15px'>  The minimum Android version required to run the app</span></a>

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#501ee6>Important things to know:</font></h2>

<a><span style='color:#501ee6; font-size:15px'>**Rows vs Columns :** |</span><span style='color:#01181c; font-size:15px'> To show all the columns and rows then follow the below code</span></a> 

<a><span style='color:#501ee6; font-size:15px'>**Warnings :** |</span><span style='color:#01181c; font-size:15px'> To ingor all the warning or hide them</span></a> 

In [4]:
# set options to be maximum for rows and columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Hide all warnings
import warnings
warnings.filterwarnings("ignore")   

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#501ee6>Data Exploration:</font></h2>

<a><span style='color:#501ee6; font-size:15px'>**Columns names:** |</span><span style='color:#01181c; font-size:15px'> To print all the column names</span></a> \
<a><span style='color:#501ee6; font-size:15px'>**Column & Rows Number:** |</span><span style='color:#01181c; font-size:15px'> To show how many rows and columns in the dataset</span></a> \
<a><span style='color:#501ee6; font-size:15px'>**Data Duplicates:** |</span><span style='color:#01181c; font-size:15px'> To show how many duplicates rows in the dataset are</span></a> \
<a><span style='color:#501ee6; font-size:15px'>**Data Information:** |</span><span style='color:#01181c; font-size:15px'> To show how many columns are, and their types</span></a> \
<a><span style='color:#501ee6; font-size:15px'>**Data Summary:** |</span><span style='color:#01181c; font-size:15px'> This will show the mean, median, quartiles, but for numericals colmns only</span></a> 



In [5]:
print(f"Show all the column names:, {df.columns}")

Show all the column names:, Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')


In [6]:
print(f"The number of rows are {df.shape[0]}, & The number of columns are {df.shape[1]}")

The number of rows are 10840, & The number of columns are 13


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h3 align="left"><font color=#501ee6>Data Exploration & Cleaning:</font></h3>

<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>It's crucial to remove them to make sure we understand the data properly.</li>
     <li>If you keep duplicates in your data, it can mess up your understanding and analysis.</li>
     <li>Duplicates can make things like averages, middle values, and spread look wrong, and they might make some data appear more important than it really is.</li>
     <li>So, it's really important to get rid of duplicates to make sure your data analysis is accurate and trustworthy.
</li></ul></span></a> 

In [7]:
# find duplicate if any
df.duplicated().sum()

483

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explain:</font></h4>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>We get over all 483 duplicates in the dataset are, So, we will remove these duplicates values.</li>

In [8]:
# remove the duplicate values
df.drop_duplicates(inplace=True)

In [9]:
# print the number of rows and columns after removing the duplicates
print(F"The number of rows in the dataseta are {df.shape[0]} & columns are:{df.shape[1]}")

The number of rows in the dataseta are 10357 & columns are:13


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explain:</font></h4>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>Check the data shape after removing all the duplicates. We get <b>10357 rows</b> & <b>13 columns</b> Now our data is clean from all the duplicates. now we can do further exploration</li>

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Data information:</font></h4>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>This will give us the information regarding the dataset</li>

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10357 entries, 0 to 10839
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10357 non-null  object 
 1   Category        10357 non-null  object 
 2   Rating          8892 non-null   float64
 3   Reviews         10357 non-null  int64  
 4   Size            10357 non-null  object 
 5   Installs        10357 non-null  object 
 6   Type            10356 non-null  object 
 7   Price           10357 non-null  object 
 8   Content Rating  10357 non-null  object 
 9   Genres          10357 non-null  object 
 10  Last Updated    10357 non-null  object 
 11  Current Ver     10349 non-null  object 
 12  Android Ver     10355 non-null  object 
dtypes: float64(1), int64(1), object(11)
memory usage: 1.1+ MB


In [11]:
df.describe()

Rating       Reviews
count  8892.000000  1.035700e+04
mean      4.187877  4.059046e+05
std       0.522377  2.696778e+06
min       1.000000  0.000000e+00
25%       4.000000  3.200000e+01
50%       4.300000  1.680000e+03
75%       4.500000  4.641600e+04
max       5.000000  7.815831e+07

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explanation:</font></h4>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
<li>If you see in the dataset there are many column with numeric values e.g., Size, Installs, Price,etc but they show as object, the reason is that they have some symbols inside which made them object instead of numeric. So, we will convert these columns into numeric ones and will check the data summary again.</li>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
<li>We will convert first <b>Size</b>, <b>Installs</b>, & <b>Price</b> respectively.</li>

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h2 align="left"><font color=#501ee6>Size Column Wrangling:</font></h2>

In [12]:
sample_size = 10
sampled_data = df.sample(n=sample_size)

# Extract and display all unique values from the "Size" column
unique_sizes = sampled_data['Size'].unique()
print(unique_sizes)


['Varies with device' '5.7M' '33M' '8.6M' '1.9M' '21M' '1.5M' '4.2M'
 '9.2M']


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explanation:</font></h4>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>If you see there are three types of values insides <B>"Size"</B> column, <b>"varies with device"</b>,<b>"M"</b>,& <b>"k"</b> .</li>


In [13]:
# Check missing values in "Size" column
df.isnull().sum()

App                  0
Category             0
Rating            1465
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          8
Android Ver          2
dtype: int64

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explanation:</font></h4>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
    <li>No missing values were found in  <B>"Size"</B> column,so we are good to go.</li>
     <li>But first we need to check how many values each category has in Size column.?</li>
    </ul></span></a> 


In [14]:
# Count rows with 'M' in the column
print(f"How many rows with 'M': {df['Size'].loc[df['Size'].str.contains('M')].value_counts().sum()}")

# Count rows with 'k' in the column
print(f"How many rows with 'k': {df['Size'].loc[df['Size'].str.contains('k')].value_counts().sum()}")

# Count rows with 'Varies with device' in the column
print(f"'Varies with device': {df['Size'].loc[df['Size'].str.contains('Varies with device')].value_counts().sum()}")


How many rows with 'M': 8516
How many rows with 'k': 315
'Varies with device': 1526


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explanation:</font></h4>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li><b>"M"</b> has <b>8516</b> values</li>
        <li><b>"k"</b> has <b>315</b> values</li>
        <li><b>"Varies with device"</b> has <b>1526</b> values</li>
    </ul></span></a> 
<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Size column strategy to clarify the column:</font></h4>
    <li>Firstly convert the whole <b>Size</b> column into bytes</li>
    <li>Check the column convert into bytes?</li>
    <li>Rename the <b>Size</b> column into <b>Size_in_Mb</b></li>
      <li>drop the  <b>Size</b> column from the data</b></li>
    </ul></span></a> 

In [15]:
# convert the whole "Size" column into bytes

# Lets define a function
def convert_size(size):
    if isinstance(size, str):
        if 'k' in size:
            return float(size.replace('k', ""))*1024
        elif 'M' in size:
            return float(size.replace('M', "")) *1024 *1024
        elif 'Varies with device' in size:
            return np.nan
    return size

In [16]:
# lets apply this function
df['Size'] = df['Size'].apply(convert_size)

In [17]:
# now make a new column that contain the "size in MB'
df['Size_in_Mb'] = df['Size'].apply(lambda x: x/(1024*1024))

In [18]:
# drop the "size_in_bytes" column
df = df.drop("Size", axis=1)
df.head(3)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   

   Reviews    Installs  Type Price Content Rating                     Genres  \
0      159     10,000+  Free     0       Everyone               Art & Design   
1      967    500,000+  Free     0       Everyone  Art & Design;Pretend Play   
2    87510  5,000,000+  Free     0       Everyone               Art & Design   

       Last Updated Current Ver   Android Ver  Size_in_Mb  
0   January 7, 2018       1.0.0  4.0.3 and up        19.0  
1  January 15, 2018       2.0.0  4.0.3 and up        14.0  
2    August 1, 2018       1.2.4  4.0.3 and up         8.7

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Column Install Wrangling:</font></h4>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>Check how many values with each installs</li>
        <li>Calculate, min, max & mean</li>
        <li>Making a new column called 'Installs_category' which will have the category of the installs</li>
        <li>Lets counts how may each category has</li>
        <li>Remove "+", "," and convert them into Integer</li>
        <li>Check again to confirm</li>
    </ul></span></a> 

In [19]:
df['Installs'].value_counts()

Installs
1,000,000+        1488
10,000,000+       1132
100,000+          1129
10,000+           1033
1,000+             890
100+               710
5,000,000+         683
500,000+           517
50,000+            474
5,000+             469
10+                385
100,000,000+       369
500+               328
50,000,000+        272
50+                204
5+                  82
1+                  67
500,000,000+        61
1,000,000,000+      49
0+                  14
0                    1
Name: count, dtype: int64

In [20]:
# Check Min and maximun no of installs
print(f"Min Installs are {df['Installs'].min()}")
print(f"Max Installs are {df['Installs'].max()}")

Min Installs are 0
Max Installs are 500,000,000+


In [21]:
# remove '+' sign
df['Installs'] = df['Installs'].apply(lambda x: x.replace('+', "") if '+' in str(x) else x)
# remove ',' sign
df['Installs'] = df['Installs'].apply(lambda x: x.replace(',', "") if ',' in str(x) else x)
# convert 'Install' column into int
df['Installs'] = df['Installs'].apply(lambda x: int(x))
# print install column type
print(f"Install column type is {df['Installs'].dtype}")

Install column type is int64


In [22]:
# making a new column called 'Installs_category' which will have the category of the installs
bins = [-1, 0, 10, 1000, 10000, 100000, 1000000, 10000000, 10000000000]
labels = ['None', 'Extremely Low', 'Minimal', 'Medium', 'Above Medium', 
          'Significant', 'Exceptionally High', 'Outstanding']
df['Install_category'] = pd.cut(df['Installs'], bins=bins, labels=labels)

In [23]:
# check to cofirm
df['Installs'].value_counts()

Installs
1000000       1488
10000000      1132
100000        1129
10000         1033
1000           890
100            710
5000000        683
500000         517
50000          474
5000           469
10             385
100000000      369
500            328
50000000       272
50             204
5               82
1               67
500000000       61
1000000000      49
0               15
Name: count, dtype: int64

In [24]:
# check to cofirm how many installs per category
df['Install_category'].value_counts()

Install_category
Minimal               2132
Significant           2005
Exceptionally High    1815
Above Medium          1603
Medium                1502
Outstanding            751
Extremely Low          534
None                    15
Name: count, dtype: int64

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explanation:</font></h4>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>Check how many values with <b>dollar</b> sign</li>
        <li>Lets counts the number of values in the "Price" column which contains 0 but doesnt contain <b>dollar</b> sign </li>
        <li>Remove the <b>dollar</b> sign</li>
        <li>Convert the column into float</li>
        <li>Check min, max and mean price</li>
        <li>Check again data summary</li>
    </ul></span></a> 

In [25]:
df['Price'].value_counts().head(5)

Price
0        9592
$0.99     146
$2.99     125
$1.99      73
$4.99      70
Name: count, dtype: int64

In [26]:
# First see how many values have $ sign
df['Price'].loc[df['Price'].str.contains('\$')].value_counts().sum()

765

In [27]:
# Lets counts the number of values in the "Price" column which contains 0 but doesnt contain $ sign
df['Price'].loc[(df['Price'].str.contains('0')) & (~df['Price'].str.contains('\$'))].value_counts().sum()

9592

In [28]:
# now we will remove the $ sign where it exists
df['Price'] = df['Price'].apply(lambda x: x.replace('$', "") if '$' in str(x) else x)
# Now we will convert the price column into float
df['Price'] = df['Price'].apply(lambda x: float(x))
df['Price'].value_counts().head(5)

Price
0.00    9592
0.99     146
2.99     125
1.99      73
4.99      70
Name: count, dtype: int64

In [29]:
# using f string print min, max and mean prices of the app
print(f"Min price is {df['Price'].min()}")
print(f"Max price is {df['Price'].max()}")
print(f"Average price is {df['Price'].mean()}")

Min price is 0.0
Max price is 400.0
Average price is 1.030800424833446


In [30]:
# check data summary
df.describe()

Rating       Reviews      Installs         Price   Size_in_Mb
count  8892.000000  1.035700e+04  1.035700e+04  10357.000000  8831.000000
mean      4.187877  4.059046e+05  1.415776e+07      1.030800    21.287413
std       0.522377  2.696778e+06  8.023955e+07     16.278625    22.540591
min       1.000000  0.000000e+00  0.000000e+00      0.000000     0.008301
25%       4.000000  3.200000e+01  1.000000e+03      0.000000     4.700000
50%       4.300000  1.680000e+03  1.000000e+05      0.000000    13.000000
75%       4.500000  4.641600e+04  1.000000e+06      0.000000    29.000000
max       5.000000  7.815831e+07  1.000000e+09    400.000000   100.000000

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h4 align="left"><font color=#501ee6>Explanation:</font></h4>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>Now see five column is showing before only two were showing. see above</li>

<a id="load_dataset"></a>
# <b><span style='color:#cc0a27'>Step: 2 |</span><span style='color:#0aaccc'> Dealing missing values</span></b>
<a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>Addressing missing values constitutes a critical aspect of the data wrangling process, as it is indispensable for obtaining accurate insights from the dataset.</li>

In [31]:
# check missing values
Missing = pd.DataFrame({'Null Values' : df.isna().sum().sort_values(ascending=False), 
                        'Percentage Null Values' : (df.isna().sum().sort_values(ascending=False)) / 
                        (df.shape[0]) * (100)})
Missing

Null Values  Percentage Null Values
Size_in_Mb               1526               14.733996
Rating                   1465               14.145023
Current Ver                 8                0.077242
Android Ver                 2                0.019311
Type                        1                0.009655
App                         0                0.000000
Category                    0                0.000000
Reviews                     0                0.000000
Installs                    0                0.000000
Price                       0                0.000000
Content Rating              0                0.000000
Genres                      0                0.000000
Last Updated                0                0.000000
Install_category            0                0.000000

In [32]:
# Totla missing values in the whole data
print(f"The total number of missing values are in data is {(df.isnull().sum().sum())}")
# And there percentage
print(f"The total number of missing values % in data is {(df.isnull().sum().sum()/len(df)*100)}")

The total number of missing values are in data is 3002
The total number of missing values % in data is 28.985227382446656


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left">

<h3 align="left"><font color=#501ee6>Observation:</font></h3>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>As we can see that in our whole data there are 3002, which is 28.98% of the whole data, The most missing values were observed in the `size_in_Mb` and `Rating' column with 1526 (14.73%), & 1465 (14.14%), respectively. While the lowest was observed in `current Ver`, `Android Ver` & `Type` with 8 (0.07%), 2 (0.01%) & 1 (0.009%) respectively.</li>
        <li>So we will remove the missing values form the `Current Ver`, `Android Ver` & `Type` because they have very less missing values and we can expect it will not affect on the actual data</li>

In [33]:
# length before removing the null values
print(f"The length of the df before removing the null values: {len(df)}")

The length of the df before removing the null values: 10357


In [34]:
# remove the rows having null values in the 'current ver', 'Android ver', & 'Type'.
df.dropna(subset=['Current Ver', 'Android Ver', 'Type'], inplace=True)

In [35]:
# length after removing the null values
print(f"The length of the df after removing the null values: {len(df)}")

The length of the df after removing the null values: 10346


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left;width:800px">

<h3 align="left"><font color=#501ee6>Explain:</font></h3>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>So in total we have removed 11 missing values from <b>Current Ver</b>, <b>Android Ver</b> & <b>Type</b> columns </li>

In [36]:
# make missing values plot by using plotly
import plotly.express as px
fig = px.bar(x=df.isnull().mean().index, y=df.isnull().sum().values)
fig.update_layout(width=650, height=400, xaxis_tickangle = -90)
fig.show()

In [37]:
# check missing values
Missing = pd.DataFrame({'Null Values' : df.isna().sum().sort_values(ascending=False), 
                        'Percentage Null Values' : (df.isna().sum().sort_values(ascending=False)) 
                        / (df.shape[0]) * (100)})
Missing

Null Values  Percentage Null Values
Size_in_Mb               1525               14.739996
Rating                   1460               14.111734
App                         0                0.000000
Category                    0                0.000000
Reviews                     0                0.000000
Installs                    0                0.000000
Type                        0                0.000000
Price                       0                0.000000
Content Rating              0                0.000000
Genres                      0                0.000000
Last Updated                0                0.000000
Current Ver                 0                0.000000
Android Ver                 0                0.000000
Install_category            0                0.000000

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left;width:800px">

<h3 align="left"><font color=#501ee6>Correlation observation:</font></h3>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>Currently, within our dataset, the sole columns that still bear missing values are 'Rating' and 'Size_in_Mb.' 

Regarding the 'Rating' column, we must exercise vigilance in addressing its missing values due to its direct linkage with the 'Installs' column, an essential factor in our dataset.

Conversely, in the 'Size' columns, we have successfully managed the 'Varies with device' values by converting them into null values. At this juncture, there is no immediate necessity for imputation in these columns, as each app's size is notably distinct, rendering it impractical to predict with a high degree of accuracy.</li>

In [38]:
import plotly.graph_objects as go
import numpy as np

# Make a correlation matrix on numeric columns
numeric_cols = ["Rating", "Reviews", "Installs", "Price", "Size_in_Mb"]
correlation_matrix = df[numeric_cols].corr()

# Create a custom text for annotating the heatmap
text_values = np.around(correlation_matrix.values, decimals=2)

fig = go.Figure(data=go.Heatmap(
    z=correlation_matrix,
    x=numeric_cols,
    y=numeric_cols,
    colorscale="YlGnBu",
    colorbar=dict(title="Correlation"),
))

# Add text annotations to the heatmap
annotations = []
for i in range(len(numeric_cols)):
    for j in range(len(numeric_cols)):
        annotations.append(
            go.layout.Annotation(
                x=numeric_cols[i],
                y=numeric_cols[j],
                text=str(text_values[j][i]),
                showarrow=False,
            )
        )

fig.update_layout(
    title="Correlation Heatmap",
    width=650,
    height=450,
    annotations=annotations,
)

fig.show()


<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left;width:800px">

<h3 align="left"><font color=#501ee6>Correlation observation:</font></h3>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>In the presented correlation matrix plot, the diversity of values is visually distinguished through a spectrum of colors, each bearing its own significance. The spectrum, ranging from lighter to darker hues, imparts a nuanced representation of the correlation levels. Specifically, lighter colors denote lower correlations, while the darker shades signify higher correlations.

In the context of correlation, the concept of lower and higher correlation pertains to the degree of association between two variables. Lower correlation implies that when one variable increases, the other exhibits only a marginal increase or change. Conversely, higher correlation implies that an increase in one variable is strongly linked to a corresponding increase in the other.

Remarkably, within our correlation matrix plot, a noteworthy correlation value emerges, standing at 0.63. This correlation pertains to the relationship between the "Rating" and "Installs" columns, signifying a substantial and positive association between these variables..</li>

In [39]:
# we can calculate the correlation matrix by using pandas
df[numeric_cols].corr()

Rating   Reviews  Installs     Price  Size_in_Mb
Rating      1.000000  0.068724  0.050869 -0.022371    0.081874
Reviews     0.068724  1.000000  0.634987 -0.009424    0.237853
Installs    0.050869  0.634987  1.000000 -0.011155    0.168805
Price      -0.022371 -0.009424 -0.011155  1.000000   -0.023820
Size_in_Mb  0.081874  0.237853  0.168805 -0.023820    1.000000

## Install category with rating

In [40]:
df.isnull().sum().sort_values(ascending=False)

Size_in_Mb          1525
Rating              1460
App                    0
Category               0
Reviews                0
Installs               0
Type                   0
Price                  0
Content Rating         0
Genres                 0
Last Updated           0
Current Ver            0
Android Ver            0
Install_category       0
dtype: int64

In [41]:
# in which Install_category the Rating has NaN values
df['Install_category'].loc[df['Rating'].isnull()].value_counts()

Install_category
Minimal               874
Extremely Low         452
Medium                 86
Above Medium           31
None                   14
Significant             3
Exceptionally High      0
Outstanding             0
Name: count, dtype: int64

In [42]:
# use groupby function to find the trend of rating in each install_category
df.groupby('Install_category')['Rating'].describe()

count      mean       std  min  25%  50%  75%  max
Install_category                                                       
None                   0.0       NaN       NaN  NaN  NaN  NaN  NaN  NaN
Extremely Low         81.0  4.637037  0.845199  1.0  4.8  5.0  5.0  5.0
Minimal             1255.0  4.170040  0.829420  1.0  3.8  4.4  4.8  5.0
Medium              1413.0  4.031352  0.605581  1.4  3.8  4.2  4.5  5.0
Above Medium        1571.0  4.094653  0.504221  1.6  3.9  4.2  4.5  4.9
Significant         2000.0  4.207100  0.378591  1.8  4.0  4.3  4.5  4.9
Exceptionally High  1815.0  4.285675  0.301071  2.0  4.1  4.3  4.5  4.9
Outstanding          751.0  4.372170  0.195374  3.1  4.3  4.4  4.5  4.8

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left;width:800px">

<h3 align="left"><font color=#501ee6>Observation:</font></h3>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>The tabulated representation above, juxtaposing 'Install_category' against the 'Rating' column, reveals the presence of missing values within our defined categories in the 'Install_category' column. Remarkably, the 'Install_category' column itself initially does not exhibit any null values. However, upon amalgamation with the 'Rating' column, the emergence of numerous null values becomes evident within our newly created categories.

To address this, we can effectively eliminate these null values by replacing them with the mean rating values corresponding to each distinct 'Install_category.' This corrective step is now poised for implementation.</li>

In [43]:
Install_cat_mean = df.groupby('Install_category')['Rating'].transform('mean')

# Replace missing values with category mean values
df['Rating'].fillna(Install_cat_mean, inplace=True)

In [44]:
# Replace NaN values in 'Install_Category' column with 'no' with 0
df['Rating'].fillna(0, inplace=True)

In [45]:
# use groupby function to find the trend of rating in each install_category
df.groupby('Install_category')['Rating'].describe()

count      mean       std  min       25%       50%  \
Install_category                                                          
None                  14.0  0.000000  0.000000  0.0  0.000000  0.000000   
Extremely Low        533.0  4.637037  0.327754  1.0  4.637037  4.637037   
Minimal             2129.0  4.170040  0.636704  1.0  4.170040  4.170040   
Medium              1499.0  4.031352  0.587941  1.4  3.800000  4.100000   
Above Medium        1602.0  4.094653  0.499316  1.6  3.900000  4.200000   
Significant         2003.0  4.207100  0.378307  1.8  4.000000  4.300000   
Exceptionally High  1815.0  4.285675  0.301071  2.0  4.100000  4.300000   
Outstanding          751.0  4.372170  0.195374  3.1  4.300000  4.400000   

                         75%  max  
Install_category                   
None                0.000000  0.0  
Extremely Low       4.637037  5.0  
Minimal             4.500000  5.0  
Medium              4.400000  5.0  
Above Medium        4.400000  4.9  
Significant         4.500000  4.9  
Exceptionally High  4.500000  4.9  
Outstanding         4.500000  4.8

In [46]:
# Check again to see whether missing impute with mean or not?
df['Install_category'].loc[df['Rating'].isnull()].value_counts()

Install_category
None                  0
Extremely Low         0
Minimal               0
Medium                0
Above Medium          0
Significant           0
Exceptionally High    0
Outstanding           0
Name: count, dtype: int64

In [47]:
df.isnull().sum().sort_values(ascending=False)

Size_in_Mb          1525
App                    0
Category               0
Rating                 0
Reviews                0
Installs               0
Type                   0
Price                  0
Content Rating         0
Genres                 0
Last Updated           0
Current Ver            0
Android Ver            0
Install_category       0
dtype: int64

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left;width:800px">

<h3 align="left"><font color=#501ee6>Observation:</font></h3>
    <a><span></span><span style='color:#01181c; font-size:15px'> <ul>
        <li>Now all the missing values in each <b>Install_category</b> is impute with the mean of the Rating.And Rating column is also filled with the mean of each Install_category column. Now still we have a size_in_Mb column which we can't impute with anything because every app has its specific size. So we can keep as it is.</li>

<a id="load_dataset"></a>
# <b><span style='color:#cc0a27'>Step: 3 |</span><span style='color:#0aaccc'> Data Visualization</span></b>
<a><span></span><span style='color:#01181c; font-size:20px'> <ul>
        <li>Let's create some questions and then try to answer them with the help of visualization</li>

<div style="border-radius:10px; padding: 15px; background-color: #f2f7f6; font-size:100%; text-align:left;width:800px">

<h4 align="left"><font color=#501ee6>Questions:</font></h4>
    <li><b>Question -1:</b> What is the rating of each category by Install?</li>
    <li><b>Question -2:</b> Show top-10 categories which has the highest number of apps</li>
    <li><b>Question -3:</b> Show top-10 categories which has the highest number of Installs</li>
    <li><b>Question -4:</b> Show top-10 categories which has the highest number of Reviews</li>
    <li><b>Question -5:</b> Show top-10 categories which has the highest number of Rating</li>
    <li><b>Question -6:</b> Show top-10 categories which has the big size In MB</li>
    <li><b>Question -7:</b> Show top-10 categories which has the higher price</li>
<h4 align="left"><font color=#501ee6>Questions by Content Rating:</font></h4>
    <li><b>Question -8:</b> What is the rating of each category by Install?</li>
    <li><b>Question -9:</b> What is the rating of each category by Install?</li>
    <li><b>Question -10:</b> What is the rating of each category by Install?</li>
    <li><b>Question -11:</b> What is the rating of each category by Install?</li>
    <li><b>Question -12:</b> What is the rating of each category by Install?</li>
    <li><b>Question -13:</b> What is the rating of each category by Install?</li>
    <li><b>Question -14:</b> What is the rating of each category by Install?</li>
    <li><b>Question -15:</b> What is the rating of each category by Install?</li>
    <li><b>Question -16:</b> What is the rating of each category by Install?</li>
  

In [48]:
# Define the custom order for the Install_category
custom_category_order = ['None ', 'Extremely Low', 'Minimal', 'Medium', 
                         'Above Medium', 'Significant', 'Exceptionally High', 'Outstanding']

fig = px.box(df, x='Install_category', y=np.log10(df['Reviews']), 
             color='Install_category', category_orders={'Install_category': custom_category_order})
fig.update_layout(
    width=650,
    height=450,
    xaxis_tickangle=-90,
    plot_bgcolor='whitesmoke',
    title='Apps categories based on reviews'
)
fig.show()

In [49]:
Apps_Category = df['Category'].value_counts().reset_index()
Apps_Category.columns = ['Category', 'Count']

# Assuming 'Apps_Category' is the DataFrame with category counts
top_10_categories = Apps_Category.nlargest(10, 'Count')  # Get the top 10 categories by count

fig = px.bar(top_10_categories, x='Category', y='Count', 
             title="Top 10 Categories of the apps", color_discrete_sequence=['red'])
fig.update_traces(width=0.5)
fig.update_xaxes(categoryorder='total descending')  # Sort categories by count in ascending order
fig.update_layout(width=650, height=500,plot_bgcolor='whitesmoke',
    xaxis_title="Category", yaxis_title="Count", xaxis_tickangle = -90)


fig.show()


In [50]:
# Calculate the area square of each district and select the top 10
cat_install = df.groupby('Category')['Installs'].mean().reset_index().nlargest(10, 'Installs')

fig = px.bar(cat_install, x='Category', y='Installs',
             title='Top 10 Categories by highest number of Installs (Mean Value)',
             color_discrete_sequence=['#f57842'])
fig.update_traces(width=0.6)
fig.update_layout(width=800, height=500,plot_bgcolor='whitesmoke')
# x axis ticks in the 90 degree angle
fig.update_layout(xaxis_tickangle=-90)

fig.update_xaxes(categoryorder='total descending')  # Optional: Sort the x-axis categories
fig.show()

In [51]:
# Calculate the area square of each district and select the top 10
cat_install = df.groupby('Category')['Reviews'].mean().reset_index().nlargest(10, 'Reviews')

fig = px.bar(cat_install, x='Category', y='Reviews',
             title='Top 10 Categories by highest number of Reviews (Mean value)',
             color_discrete_sequence=['#f57842'])
fig.update_traces(width=0.6)
fig.update_layout(width=800, height=500,plot_bgcolor='whitesmoke')
# x axis ticks in the 90 degree angle
fig.update_layout(xaxis_tickangle=-90)

fig.update_xaxes(categoryorder='total descending')  # Optional: Sort the x-axis categories
fig.show()

In [52]:
# Calculate the area square of each district and select the top 10
cat_install = df.groupby('Category')['Rating'].mean().reset_index().nlargest(10, 'Rating')

fig = px.bar(cat_install, x='Category', y='Rating',
             title='Top 10 Categories by highest number of Ratings (Mean value)',
             color_discrete_sequence=['#f57842'])
fig.update_traces(width=0.6)
fig.update_layout(width=800, height=500,plot_bgcolor='snow')
# x axis ticks in the 90 degree angle
fig.update_layout(xaxis_tickangle=-90)

fig.update_xaxes(categoryorder='total descending')  # Optional: Sort the x-axis categories
fig.show()

In [53]:
# Calculate the area square of each district and select the top 10
cat_install = df.groupby('Category')['Size_in_Mb'].mean().reset_index().nlargest(10, 'Size_in_Mb')

fig = px.bar(cat_install, x='Category', y='Size_in_Mb',
             title='Top 10 Categories by big size in Mb (Mean value)',
             color_discrete_sequence=['#f57842'])
fig.update_traces(width=0.6)
fig.update_layout(width=800, height=500,plot_bgcolor='whitesmoke')
# x axis ticks in the 90 degree angle
fig.update_layout(xaxis_tickangle=-90)

fig.update_xaxes(categoryorder='total descending')  # Optional: Sort the x-axis categories
fig.show()

In [54]:
# Calculate the area square of each district and select the top 10
cat_install = df.groupby('Category')['Price'].mean().reset_index().nlargest(10, 'Price')

fig = px.bar(cat_install, x='Category', y='Price',
             title='Top 10 Categories by higher price (Mean value)',
             color_discrete_sequence=['#f57842'])
fig.update_traces(width=0.6)
fig.update_layout(width=800, height=500,plot_bgcolor='whitesmoke')
# x axis ticks in the 90 degree angle
fig.update_layout(xaxis_tickangle=-90)

fig.update_xaxes(categoryorder='total descending')  # Optional: Sort the x-axis categories
fig.show()

# How many app developed for content rating columns

In [55]:
Apps_Category = df['Content Rating'].value_counts().reset_index()
Apps_Category.columns = ['Content Rating', 'Count']

# Assuming 'Apps_Category' is the DataFrame with category counts
top_10_categories = Apps_Category.nlargest(10, 'Count')  # Get the top 10 categories by count

fig = px.bar(top_10_categories, x='Content Rating', y='Count', 
             color = 'Content Rating', title="App developed for specific age group individuals")
fig.update_traces(width=0.5)
fig.update_xaxes(categoryorder='total descending')  # Sort categories by count in ascending order
fig.update_layout(width=700, height=450,plot_bgcolor='snow',
    xaxis_title="Category", yaxis_title="Count", xaxis_tickangle = -90)


fig.show()


In [56]:
df_teens = df[df['Content Rating'] == 'Everyone']
df_teens.head(2)

# Group the filtered data by 'category' and count the occurrences
category_counts = df_teens['Category'].value_counts().reset_index()
category_counts

category_counts.columns = ['Category', 'count']

# Sort the data by count in descending order
category_counts = category_counts.sort_values(by='count', ascending=False)

# Select the top 10 categories
category_counts_top_10 = category_counts.head(10)

# Create a Plotly bar chart for the top 10 categories
fig = px.bar(category_counts_top_10, x='Category', y='count', 
             title='Top Categories liked by Everyone', color = 'Category')
# Fig layout
fig.update_traces(width=0.5)
fig.update_layout(width=800, height=500, xaxis_tickangle=-90)

# Show the plot
fig.show()

In [57]:
df_teens = df[df['Content Rating'] == 'Teen']
df_teens.head(2)

# Group the filtered data by 'category' and count the occurrences
category_counts = df_teens['Category'].value_counts().reset_index()
category_counts

category_counts.columns = ['Category', 'count']

# Sort the data by count in descending order
category_counts = category_counts.sort_values(by='count', ascending=False)

# Select the top 10 categories
category_counts_top_10 = category_counts.head(10)

# Create a Plotly bar chart for the top 10 categories
fig = px.bar(category_counts_top_10, x='Category', y='count', title='Top Categories liked by Teens', color = 'Category')
# Fig layout
fig.update_traces(width=0.5)
fig.update_layout(width=800, height=500, xaxis_tickangle=-90)

# Show the plot
fig.show()

In [58]:
df_teens = df[df['Content Rating'] == 'Mature 17+']
df_teens.head(2)

# Group the filtered data by 'category' and count the occurrences
category_counts = df_teens['Category'].value_counts().reset_index()
category_counts

category_counts.columns = ['Category', 'count']

# Sort the data by count in descending order
category_counts = category_counts.sort_values(by='count', ascending=False)

# Select the top 10 categories
category_counts_top_10 = category_counts.head(10)

# Create a Plotly bar chart for the top 10 categories
fig = px.bar(category_counts_top_10, x='Category', y='count', title='Top Categories liked by Mature 17 plus', color = 'Category')
# Fig layout
fig.update_traces(width=0.5)
fig.update_layout(width=800, height=500, xaxis_tickangle=-90)

# Show the plot
fig.show()

In [59]:
df_teens = df[df['Content Rating'] == 'Adults only 18+']
df_teens.head(2)

# Group the filtered data by 'category' and count the occurrences
category_counts = df_teens['Category'].value_counts().reset_index()
category_counts

category_counts.columns = ['Category', 'count']

# Sort the data by count in descending order
category_counts = category_counts.sort_values(by='count', ascending=False)

# Select the top 10 categories
category_counts_top_10 = category_counts.head(10)

# Create a Plotly bar chart for the top 10 categories
fig = px.bar(category_counts_top_10, x='Category', y='count', 
             title='Top Categories liked by Adults only eighteen plus', color = 'Category')
# Fig layout
fig.update_traces(width=0.5)
fig.update_layout(width=800, height=500, xaxis_tickangle=-90)

# Show the plot
fig.show()

In [60]:
fig = px.scatter(df, x=np.log10(df['Reviews']), y=np.log10(df['Installs']))
fig.update_layout(width=600, height=400,
    title="Scatter Plot of Reviews vs Installs",
    xaxis_title="Reviews",
    yaxis_title="Installs"
)
fig.show()